Add a Scroll Bar

enricozb opened the pull request:
A scroll bar that also displays selection lines in the buffer.

This PR adds:
- `terminal_scroll_bar` to `ui_options`, which controls whether the scroll bar should show.
- `ScrollBarGutter` and `ScrollBarHandle` faces describing how to style the gutter and handle. The foreground of these faces will be the color of the characters used to display selection lines.

## Screenshots

Here is a recording with the default styling and config
[![asciicast](https://asciinema.org/a/1yqgd30qQ1G4OkWvOx4FfT29y.svg)](https://asciinema.org/a/1yqgd30qQ1G4OkWvOx4FfT29y)

Here is a recording with my personal styling and config:
[![asciicast](https://asciinema.org/a/IZGDvrHV2ntDaaKrWCnqoCKFN.svg)](https://asciinema.org/a/IZGDvrHV2ntDaaKrWCnqoCKFN)

I've tested this with both non-remote and remote clients.

## Issues

I am not the biggest fan of the implementation. The computation of `Vector<LineCount> selection_lines` happens on every draw call. This is needed to serialize these locations for `RemoteUI` and `JsonUI`. This is also computed regardless if the scrollbar is enabled, which is unfortunate.

I'm not sure what the appetite for adding something like this is, especially with my implementation. I saw there was a highlighter implementation (https://github.com/mawww/kakoune/pull/2597) but did not honestly consider it. If we feel like this is best done as a highlighter I can investigate that approach as well.

enricozb committed:
add scroll bar

enricozb committed:
remove cerr

enricozb committed:
add faces and selection chars

enricozb committed:
fix remote issues

enricozb committed:
remove debug statements

enricozb committed:
empty line

enricozb committed:
resize scroll_bar_scratch on resize

Delapouite commented:
Thanks! I appreciate that it does more than a "regular" scroll bar by adding visual markers for selections.

Of course some other text-editors/IDE offer much more with other kinds of markers such as git added/changed/removed lines, or info/warning/error diagnostics. But then you end up with a "3 lanes" scroll bars with a lot of infos which can be confusing. "There's a red mark at the beginning of the buffer : is it red because of git removed lined or red because it's an LSP error?"

So here, I think you strike the right balance by only focusing on one of the core principle of Kakoune which is its multi-selections capabilities.

enricozb committed:
clamp mark_height, fix resize

enricozb committed:
more precise geometry calculations

unfortunately this method causes the scrollbar to potentially have a
non-constant height.

however this method has the nice property of scrolling the scrollbar past the
end of the buffer, just like we do with padding.

enricozb committed:
better compatibility with wrap highlighter

when the wrap highlighter is added, the scroll bar size will reflect how
many line numbers are visible. if a wrapped line is effectively taking
up a lot of the vertical space in the window, the scroll bar handle will
be smaller to reflect that the user is viewing a smaller portion of the
file relative to the number of lines in the file.

enricozb commented:
I've improved the calculation of the size of the scroll bar handle. It works well with the `wrap` highlighter. It will decrease in size if the number of lines displayed to the user is small (due to a wrapped line taking up a lot of vertical space).

There is still a bug where when scrolling horizontally (which is not possible when the `wrap` highlighter is added), the scroll bar position is totally incorrect. This is because the display buffer `range` changes when scrolling horizontally. Still trying to fix this.

enricozb committed:
move div_round_up back to its original place

enricozb committed:
fix height relative to the number of visible lines

prior to this the height of the scrollbar could change even if the
nubmer of visible lines was vixed.

enricozb committed:
fix padding_lines calculation

enricozb committed:
Merge branch 'mawww:master' into enricozb/scroll-bar

enricozb committed:
Merge branch 'master' into enricozb/scroll-bar

raiguard commented:
Also, @mawww is this something you are willing to consider merging? It is a great usability improvement.

enricozb committed:
Merge branch 'master' into enricozb/scroll-bar

enricozb committed:
style

enricozb commented:
It's been a while since I've looked at this code, but I've merged master in and fixed the conflicts. It looks like a `DisplayCoord cursor_pos` argument was added to a few functions, i've also marked it as `const` as it wasn't being modified.

enricozb commented:
This consistently happens to me on every client creation, and is fixed after a resize.

It's likely to do with some computation i wrote that isn't refreshed on changes to `terminal_scroll_bar` in `ui_options`.

enricozb committed:
change dimensions only if m_scroll_bar

enricozb committed:
Merge branch 'mawww:master' into enricozb/scroll-bar

enricozb commented:
Fixed the incorrect dimension calculation, no longer requires a resize. Also updated to master once again.

enricozb committed:
fix incorrect documentation

Yus314 commented:
I've submitted #5471 which extracts the viewport information subset from this PR as an independent protocol extension — adding `first_buffer_line`, `last_buffer_line`, and `buffer_line_count` to the JSON UI `draw` message.

This is the approach I described in my [earlier comment](https://github.com/mawww/kakoune/pull/5304#issuecomment-2705870508): splitting the protocol change from the scroll bar UI, so alternative frontends can benefit from viewport data regardless of whether a built-in scroll bar is merged.
